lab5.md

```diff
@@ -0,0 +1,63 @@
+# lab5 - SOCKSv5 Proxy
+## Lab Description
+_Proxy server of standart SOCKS (version 5) which supports one command - "establish a TCP/IP stream connection". Handling of several clients is implemented using default Java-API (see: Selector, SelectionKey, SelectableChannel etc. in Java API references)._
+
+### Proxying
+A little about proxying:
+1. Server accepts new Client
+2. Client sends to server a 'connection request' which contains dst-host address and port.
+3. Server establishes a connection to dst host.
+4. Proxying starts - server implements data traversing between dst-host and client.
+
+### SOCKSv5 protocol
+See full protocol description [here](https://www.ietf.org/rfc/rfc1928.txt) or on [wikipedia](https://ru.wikipedia.org/wiki/SOCKS). 
+1. The Client sends "Hello" message to Server, which contains protocol version (5 in our case), and numbers of supported authentication methods. 
+
+2. Server responses with message which contains protocol version and number of choosen authentication method. 
+3. Then the Client sends 'Connection request' message which contains 'command code' (0x01 - 'establish TCP/IP connection' in our case), 'dst address type', 'dst address value' and 'dst port number'. 
+4. Server established connection to dst host and sends response to the Client - with the same data as in client request except that instead of 'command code' there is 'response code' (0x00 in casse of success)
+5. Then proxying starts
+
+___
+<a href="https://imgbb.com/"><img src="https://i.ibb.co/pvF0rsH/sequence.png" alt="sequence" border="0"></a>
+___
+
```

Comment:
Один важный момент упущен - сколько потоков на сервере (один, НЯП), и как организавана обработка множественных запросов в этом одном потоке. Типа
The proxy sever runs all communication with the clients in a single thread. Client processing is organized in "rounds". In each round server obtains outstanding requests from clients, then processes them one by one. While server is busy processing clients, new requests coming from clients are buffered by the Java runtime network layer, so that next time sever gets to polling for outstanding requests, it gets what has been buffered while the server was busy. Each client must wait for its current request to be processed by server before issuing the next (или нет - это надо описать).

Reply:
исправил